std/mem.zig

```diff
@@ -320,63 +333,98 @@ pub fn writeInt(buf: []u8, value: var, endian: builtin.Endian) void {
 /// any of the bytes in `split_bytes`.
 /// split("   abc def    ghi  ", " ")
 /// Will return slices for "abc", "def", "ghi", null, in that order.
-pub fn split(buffer: []const u8, split_bytes: []const u8) SplitIterator {
-    return SplitIterator {
-        .index = 0,
-        .buffer = buffer,
-        .split_bytes = split_bytes,
-    };
-}
+pub fn split(comptime viewType: type, comptime iterator_type: type, comptime baseType: type) type {
```

Comment:
I don't think the things in `mem` should use these "string iterators". When I use `mem.split` I only care about memory, not encoding, views, interators and all that.

Reply:
The `next` fn in here is a good reason, why `mem.split` shouldn't use these iterators. Iterating over memory can't fail with an error, because the elements don't need to be validated to decoded.

Reply:
I disagree with both points because of a few reasons :).
1) We need split in a few cases (just in std); first for string splitting in windows which is Utf16 with a few weird oddities, with my example you can just create a Utf16Win view and corresponding iterator and you don't have to create a unique function for it :), we also need a generic Utf8 one for 'general use' and a raw Ascii one for memory or when you know your just dealing with bland non encoded data.

Thus if we just pretend that iterators and views don't exist it becomes combersome :).

2) You generally don't have to care about views and iterators :) as shown in the test example there exists a type under stringUtils called string_it which maps the iterators and views for you :).  That is instead of calling `split("A/B/C/", "/")` you call `stringUtils.string_it.init("A/B/C/", "/")`, currently 'next' returns a view however I'm also gonna add a function that returns the raw type that you specify which in the case of the string one is []const u8, this should solve your issue of views as you'll never need to touch one.  Views are typically a good thing as they generalise.

3) Your second statement is also false, Iterators don't EVER fail with an error :), just look at both the Utf8 Iterator and Ascii Iterator they have type ?[]u8 not !?[]u8 because of course they don't fail, HOWEVER the views can fail if you give them invalid data, I can change it to just set the data regardless however this would then ensure that the data you originally give it has to be valid which it doesn't necessarily have to be :), however technically you are right next doesn't have to be error checked since I can validate that the slice can use initUnchecked.

So overall I can make next() not return an error ever and it kinda is just due to me wanting to be 100% safe and that kinda eroded the practicality of it :).  So with all that said I don't particularly get why you don't like it?

Reply:
My only real problem is that I don't think it should be in `mem`. Memory splitting and string splitting is different. When we split memory, there is no concept of valid code points, encoding and all that. Therefore I think the interface should be simpler.

I agree, that for splitting paths, the string splitting functions you provide should be used, as paths are strings. However, there are cases where you just have some memory ([]u8, []16, []32, []64) and want to split on some values. Btw, `stringUtils.string_it` only works on ascii, so you can't split any arbitrary sequence of bytes.

Reply:
I guess you could provide a memory view and iterator though, and that might work out. I really think `split` should not be able to fail not `next` for these though :)

Reply:
Yeh it won't fail for next I already finished that, currently it can only fail upon initialisation of the iterator.  Also you can get it to return the raw data instead of a view.

And yes it only works for ascii that is just due to me not bothering naming it properly.  I think I'll name it stringUtils.split_ascii_it, and have stringUtils.split_it for Utf8 and have stringUtils.split_win_it for the windows special Utf16 (that is super annoying why can't windows do anything right lel).

Again I completely agree next will not fail for anything after my next set of changes :).

One last thing; so you would prefer mem.split to use the old system and have the new split exist under stringUtils :)?

Really appreciate you having an in-depth look at my code by the way, your points are really valid and are helpful 😄 .

Reply:
I think so. I think `mem` should only concern itself with slices of data. Not a more complicated iterator/view system. Ofc, I don't have the final say, as I'm not a "designer" of Zig or it's std library.

Reply:
One note: paths are not strings in linux. They are null-terminated arrays of bytes. You can have invalid utf-8 as a linux file path.

Reply:
Oh yes I'll make sure that it allows for null-terminated arrays of bytes :).  Pretty sure that path sanitises it to a slice anyway but I'll double check.